packages/core/core/src/factories/useUserBillingFactory.ts

```diff
@@ -0,0 +1,128 @@
+import { Ref, ComputedRef, computed } from '@vue/composition-api';
+import { UseUserBilling } from '../types';
+import { sharedRef } from '../utils';
+
+export interface UseUserBillingFactoryParams<ADDRESS> {
+  addAddress: (params: {
+    address: Readonly<ADDRESS>;
+    addresses: Readonly<ComputedRef<ADDRESS[]>>;
+  }) => Promise<ADDRESS[]>;
+  deleteAddress: (params: {
+    address: Readonly<ADDRESS>;
+    defaultAddress: Readonly<ComputedRef<ADDRESS>>;
+    addresses: Readonly<ComputedRef<ADDRESS[]>>;
+  }) => Promise<ADDRESS[]>;
+  updateAddress: (params: {
+    address: Readonly<ADDRESS>;
+    defaultAddress: Readonly<ComputedRef<ADDRESS>>;
+    addresses: Readonly<ComputedRef<ADDRESS[]>>;
+  }) => Promise<ADDRESS[]>;
+  load: (params: {
+    addresses: Readonly<ComputedRef<ADDRESS[]>>;
+  }) => Promise<ADDRESS[]>;
+  setDefault: (params: {
+    address: Readonly<ADDRESS>;
+    defaultAddress: Readonly<ComputedRef<ADDRESS>>;
+    addresses: Readonly<ComputedRef<ADDRESS[]>>;
+  }) => Promise<ADDRESS>;
+}
+
+export const useUserBillingFactory = <ADDRESS>(
+  factoryParams: UseUserBillingFactoryParams<ADDRESS>
+) => {
+
+  const useUserBilling = (): UseUserBilling<ADDRESS> => {
+    const defaultAddress: Ref<ADDRESS> = sharedRef(null, 'useUserBilling-default-address');
+    const loading: Ref<boolean> = sharedRef(false, 'useUserBilling-loading');
+    const addresses: Ref<ADDRESS[]> = sharedRef([], 'useUserBilling-addresses');
+
+    const readonlyAddresses: ComputedRef<ADDRESS[]> = computed(() => addresses.value);
```

Comment:
using typescript `readonly` in functions seems more sufficient for this job

Reply:
```ts
const readonlyDefaultAddress: Readonly<ADDRESS> = unref(defaultAddress);
```